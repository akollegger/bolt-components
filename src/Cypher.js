import React, { Component } from "react";
import PropTypes from "prop-types";

export const missingRenderPropError =
  "Pass render property to render something. <Cypher render={({pending, error, response} => {})} />";
export const missingQueryError =
  "A property named 'query' is missing. Component needs a query to be executed.";
export const missingDriverError =
  "A property or conext object named 'driver' is missing. Component needs a connected driver to send queries over. See <Provider>.";

/**
 * Interface
 * <Cypher
 *  render = fn to be invoked on any state change in the component
 *  query = the query to be executed
 *  params = optional params to send with the query
 *  interval = run query at an interval in seconds
 *  driver = connected neo4j-driver
 * />
 */

class Cypher extends Component {
  constructor(props, context) {
    super(props, context);
    this.driver = props.driver || context.driver;
  }
  state = {
    result: null,
    pending: true,
    error: null,
    tick: 0
  };
  componentDidMount() {
    this.query(this.props);
    if (this.props.interval) {
      this.interval = setInterval(
        () => this.query(this.props),
        this.props.interval * 1000
      );
    }
  }
  componentWillUnmount() {
    clearInterval(this.interval);
  }
  query(props) {
    const { query, params = null } = props;
    if (!query) throw new Error(missingQueryError);
    if (!this.driver) throw new Error(missingDriverError);
    this.setState({ pending: true }, () => {
      const session = this.driver.session();
      session
        .run(query, params)
        .then(res => {
          this.setState({
            pending: false,
            result: res,
            error: null,
            tick: this.state.tick + 1
          });
        })
        .catch(error => {
          this.setState({
            pending: false,
            result: null,
            error: error,
            tick: this.state.tick + 1
          });
        });
    });
  }
  render() {
    if (!this.props.render) {
      throw new Error(missingRenderPropError);
      return;
    }
    return this.props.render({ ...this.state });
  }
}

Cypher.contextTypes = {
  driver: PropTypes.object
};

export default Cypher;
