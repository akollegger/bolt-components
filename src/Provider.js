import { Component } from "react";
import PropTypes from "prop-types";

class Provider extends Component {
  constructor(props, context) {
    super(props, context);
  }

  getChildContext() {
    return {
      driver: this.props.driver
    };
  }

  render() {
    const children = this.props.children;
    return Array.isArray(children) ? children[0] : children;
  }
}

Provider.childContextTypes = {
  driver: PropTypes.object
};

Provider.propTypes = {
  driver: PropTypes.object.isRequired,
  children: PropTypes.node.isRequired
};

export default Provider;
