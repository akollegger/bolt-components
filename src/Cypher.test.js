import React from "react";
import TestRenderer from "react-test-renderer";
import { mockDriver } from "../config/test_helpers";
import Cypher, {
  missingRenderPropError,
  missingQueryError,
  missingDriverError
} from "./Cypher";
import Provider from "./Provider";

it("renders without crashing", () => {
  const out = TestRenderer.create(
    <Cypher query="x" driver={mockDriver()} render={() => null} />
  );
  expect(out.toJSON()).toEqual(null);
});

it("throws on missing render prop", () => {
  expect(() => {
    TestRenderer.create(<Cypher />);
  }).toThrowError(missingRenderPropError);
});

it("throws on missing query prop", () => {
  expect(() => {
    TestRenderer.create(<Cypher render={() => null} />);
  }).toThrowError(missingQueryError);
});

it("throws on missing driver prop and driver context", () => {
  expect(() => {
    TestRenderer.create(<Cypher query="x" render={() => null} />);
  }).toThrowError(missingDriverError);
});

it("renders with driver as a prop", () => {
  const out = TestRenderer.create(
    <Cypher driver={mockDriver()} query="x" render={() => null} />
  );
  expect(out.toJSON()).toEqual(null);
});

it("renders with driver context", () => {
  const out = TestRenderer.create(
    <Provider driver={mockDriver()}>
      <Cypher query="x" render={() => null} />
    </Provider>
  );
  expect(out.toJSON()).toEqual(null);
});

it("runs cypher query on mount and not on new props", () => {
  // Given
  const spy = jest.fn((query, params) => Promise.resolve());
  const query = "RETURN rand()";

  // When
  const r = TestRenderer.create(
    <Provider driver={mockDriver(spy)}>
      <Cypher query={query} render={() => null} />
    </Provider>
  );

  // Then
  expect(spy).toHaveBeenLastCalledWith(query, undefined);
  expect(spy).toHaveBeenCalledTimes(1);

  // When
  r.update(
    <Provider driver={mockDriver(spy)}>
      <Cypher query={"xxx"} render={() => null} />
    </Provider>
  );

  // Then
  expect(spy).toHaveBeenLastCalledWith(query, undefined);
  expect(spy).toHaveBeenCalledTimes(1);
});

it("runs cypher query at an interval", () => {
  // Given
  jest.useFakeTimers();

  const spy = jest.fn((query, params) => Promise.resolve());
  const query = "RETURN rand()";

  // When
  const r = TestRenderer.create(
    <Provider driver={mockDriver(spy)}>
      <Cypher
        interval={1} // every second
        query={query}
        render={() => null}
      />
    </Provider>
  );

  // Then
  expect(spy).toHaveBeenLastCalledWith(query, undefined);
  expect(spy).toHaveBeenCalledTimes(1);

  // When
  jest.runOnlyPendingTimers();

  // Then
  expect(spy).toHaveBeenLastCalledWith(query, undefined);
  expect(spy).toHaveBeenCalledTimes(2);

  // When
  jest.runOnlyPendingTimers();

  // Then
  expect(spy).toHaveBeenLastCalledWith(query, undefined);
  expect(spy).toHaveBeenCalledTimes(3);
});

it("passes result argument to render function", () => {
  // Given

  const renderSpy = jest.fn();
  renderSpy.mockImplementation(({ pending, error, result }) => {
    return pending ? "pending" : error ? error : result;
  });
  const closeSpy = jest.fn();
  const runSpy = jest.fn();
  runSpy.mockImplementation(() => Promise.resolve(10));
  const query = "CALL mock";

  // When
  const r = TestRenderer.create(
    <Provider driver={mockDriver(runSpy, closeSpy)}>
      <Cypher query={query} render={renderSpy} />
    </Provider>
  );

  // Then
  let tree = r.toJSON();
  expect(runSpy).toHaveBeenLastCalledWith(query, undefined);
  expect(runSpy).toHaveBeenCalledTimes(1);
  expect(closeSpy).toHaveBeenCalledTimes(0);
  expect(renderSpy).toHaveBeenLastCalledWith({
    pending: true,
    error: null,
    result: null,
    tick: 0
  });
  expect(renderSpy).toHaveBeenCalledTimes(2); // Initial render + pending
  expect(tree).toMatchSnapshot();

  // This is needed to flush the promise chain
  return flushPromises().then(() => {
    tree = r.toJSON();
    expect(closeSpy).toHaveBeenCalledTimes(1);
    expect(renderSpy).toHaveBeenLastCalledWith({
      pending: false,
      error: null,
      result: 10,
      tick: 1
    });
    expect(renderSpy).toHaveBeenCalledTimes(3);
    expect(tree).toMatchSnapshot();
  });
});

it("passes error argument to render function", () => {
  // Given

  const renderSpy = jest.fn();
  renderSpy.mockImplementation(({ pending, error, result }) => {
    return pending ? "pending" : error ? error : result;
  });
  const runSpy = jest.fn();
  runSpy.mockImplementation(() => Promise.reject("ERROR"));
  const closeSpy = jest.fn();
  const query = "CALL mock";

  // When
  const r = TestRenderer.create(
    <Provider driver={mockDriver(runSpy, closeSpy)}>
      <Cypher query={query} render={renderSpy} />
    </Provider>
  );

  // Then
  let tree = r.toJSON();
  expect(runSpy).toHaveBeenLastCalledWith(query, undefined);
  expect(runSpy).toHaveBeenCalledTimes(1);
  expect(closeSpy).toHaveBeenCalledTimes(0);
  expect(renderSpy).toHaveBeenCalledTimes(2); // Initial render + pending
  expect(tree).toMatchSnapshot();

  // This is needed to flush the promise chain
  return flushPromises().then(() => {
    tree = r.toJSON();
    expect(closeSpy).toHaveBeenCalledTimes(0); // no closing sessions on error
    expect(renderSpy).toHaveBeenLastCalledWith({
      pending: false,
      error: "ERROR",
      result: null,
      tick: 1
    });
    expect(renderSpy).toHaveBeenCalledTimes(3);
    expect(tree).toMatchSnapshot();
  });
});

function flushPromises() {
  return new Promise(resolve => setImmediate(resolve));
}
