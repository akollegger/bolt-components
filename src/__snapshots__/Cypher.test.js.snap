// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes error argument to render function 1`] = `"pending"`;

exports[`passes error argument to render function 2`] = `"ERROR"`;

exports[`passes result argument to render function 1`] = `"pending"`;

exports[`passes result argument to render function 2`] = `"10"`;
